File: benchmarks/values/float_mandelbrot.c
#include <stdio.h>
#include <time.h>

int main(void) {
    struct timespec t0, t1;
    clock_gettime(CLOCK_MONOTONIC, &t0);

    long long total = 0;
    for (int py = 0; py < 2000; py++) {
        double ci = -1.5 + py * 3.0 / 2000.0;
        for (int px = 0; px < 2000; px++) {
            double cr = -2.0 + px * 3.0 / 2000.0;
            double zr = 0.0, zi = 0.0;
            int iter = 0;
            while (iter < 50) {
                if (zr * zr + zi * zi > 4.0) break;
                double new_zr = zr * zr - zi * zi + cr;
                zi = 2.0 * zr * zi + ci;
                zr = new_zr;
                iter++;
            }
            total += iter;
        }
    }

    clock_gettime(CLOCK_MONOTONIC, &t1);
    double elapsed = (t1.tv_sec - t0.tv_sec) + (t1.tv_nsec - t0.tv_nsec) / 1e9;
    printf("%lld\n", total);
    printf("elapsed: %.3fs\n", elapsed);
    return 0;
}

File: benchmarks/values/float_mandelbrot.rs
use std::time::Instant;

fn main() {
    let t0 = Instant::now();

    let mut total: i64 = 0;
    for py in 0..2000 {
        let ci = -1.5 + py as f64 * 3.0 / 2000.0;
        for px in 0..2000 {
            let cr = -2.0 + px as f64 * 3.0 / 2000.0;
            let mut zr = 0.0_f64;
            let mut zi = 0.0_f64;
            let mut iter: i64 = 0;
            while iter < 50 {
                if zr * zr + zi * zi > 4.0 {
                    break;
                }
                let new_zr = zr * zr - zi * zi + cr;
                zi = 2.0 * zr * zi + ci;
                zr = new_zr;
                iter += 1;
            }
            total += iter;
        }
    }

    let elapsed = t0.elapsed();
    println!("{}", total);
    println!("elapsed: {:.3}s", elapsed.as_secs_f64());
}
